rust/core-lib/src/fuchsia/sync.rs

```diff
@@ -0,0 +1,186 @@
+    }
+
+    /// Called whenever this app changed its own state and would like to
+    /// persist the changes to the ledger. Changes can't be committed
+    /// immediately since we have to wait for PageWatcher changes that may not
+    /// have arrived yet.
+    pub fn state_changed(&mut self) {
+        if !self.transaction_pending {
+            self.transaction_pending = true;
+            let ready_future = self.page.start_transaction();
+            let done_chan = self.updates.clone();
+            let buffer = self.buffer.clone();
+            ready_future.with(move |res| {
+                assert_eq!(Status_Ok, res.unwrap(), "failed to start transaction");
+                done_chan.send(SyncMsg::TransactionReady { buffer }).unwrap();
+            });
+        }
+    }
+
+    /// Should be called in SyncContainer::transaction_ready to persist the current state.
+    pub fn commit_transaction(&mut self, state: &Engine) {
+        assert!(self.transaction_pending, "must call state_changed (and wait) before commit");
+        self.page.put(self.key.clone(), state_to_buf(state)).with(ledger_crash_callback);
+        self.page.commit().with(ledger_crash_callback);
+        self.transaction_pending = false;
+    }
+}
+
+/// All the different asynchronous events the updater thread needs to listen for and act on
+pub enum SyncMsg {
+    NewState { buffer: BufferIdentifier, new_buf: Vec<u8>, done: Option<Promise<Option<PageSnapshot_Server>, fidl::Error>> },
+    TransactionReady { buffer: BufferIdentifier },
+    /// Shut down the updater thread
+    Stop
+}
+
+/// We want to be able to register to recieve events from inside the
+/// `SyncStore`/`SyncContainer` but from there we don't have access to the
+/// Mutex that holds the container, so we give channel Senders to all the
+/// futures so that they can all trigger events in one place that does have
+/// the right reference.
+///
+/// Additionally, the individual `Editor`s aren't wrapped in a `Mutex` so we
+/// have to hold a `BufferContainerRef` and use `BufferIdentifier`s with one
+/// `SyncUpdater` for all buffers.
+pub struct SyncUpdater<W: Write> {
+    container_ref: BufferContainerRef<W>,
+    chan: Receiver<SyncMsg>,
+}
+
+impl<W: Write + Send + 'static> SyncUpdater<W> {
+    pub fn new(container_ref: BufferContainerRef<W>, chan: Receiver<SyncMsg>) -> SyncUpdater<W> {
+        SyncUpdater { container_ref, chan }
+    }
+
+    /// Run this in a thread, it will return when it encounters an error
+    /// reading the channel or when the `Stop` message is recieved.
+    pub fn work(&self) -> Result<(),RecvError> {
+        loop {
+            let msg = self.chan.recv()?;
+            match msg {
+                SyncMsg::Stop => return Ok(()),
+                SyncMsg::TransactionReady { buffer }=> {
+                    let mut container = self.container_ref.lock();
+                    let mut editor = container.editor_for_buffer_mut(&buffer).unwrap();
```

Comment:
Yeah, agree this looks like it can happen in normal use and shouldn't be a panic.